sdk/go/pulumi-language-go/main.go

```diff
@@ -734,10 +755,143 @@ func runCmdStatus(cmd *exec.Cmd, env []string) (int, error) {
+		return err
+	}
+	defer func() {
+		contract.IgnoreError(t.Stop())
+		t.Cleanup()
+	}()
+	for line := range t.Lines {
+		if line.Err != nil {
+			continue
+		}
+		if addr, ok := strings.CutPrefix(line.Text, "API server listening at: "); ok {
+			host, port, err := net.SplitHostPort(addr)
+			if err != nil {
+				return err
+			}
+			c.Host = host
+			p, err := strconv.ParseInt(port, 10, 32)
+			if err != nil {
+				return err
+			}
+			c.Port = p
+			break
+		}
+	}
+	return nil
+}
+
+func (c *debugger) Cleanup() {
+	contract.IgnoreError(os.Remove(c.LogDest))
+}
+
+func debugCommand(bin string) (*exec.Cmd, *debugger, error) {
+	godlv, err := executable.FindExecutable("dlv")
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to find 'dlv' executable: %w", err)
+	}
+	logFile, err := os.CreateTemp("", "pulumi-go-dlv-")
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to allocate tmp file: %w", err)
+	}
+	contract.IgnoreClose(logFile)
+	args := []string{"--headless=true", "--api-version=2"}
+	args = append(args, "--log", "--log-dest", logFile.Name())
+	args = append(args, "exec", bin)
+	dlvCmd := exec.Command(godlv, args...)
+	return dlvCmd, &debugger{Host: "127.0.0.1", LogDest: logFile.Name()}, nil
+}
+
+func startDebugging(ctx context.Context, engineClient pulumirpc.EngineClient, cmd *exec.Cmd, dbg *debugger) error {
+	// wait for the debugger to be ready
+	ctx, _ = context.WithTimeoutCause(ctx, 1*time.Minute, errors.New("debugger startup timed out"))
+	err := dbg.WaitForReady(ctx)
+	if err != nil {
+		return err
+	}
+
+	debugConfig, err := structpb.NewStruct(map[string]interface{}{
+		"name":    "Pulumi: Program (Go)",
+		"type":    "go",
+		"request": "attach",
+		"mode":    "remote",
+		"host":    dbg.Host,
+		"port":    dbg.Port,
+	})
+	if err != nil {
+		return err
+	}
+	_, err = engineClient.StartDebugger(ctx, &pulumirpc.StartDebuggerRequest{
+		Config:  debugConfig,
+		Message: "on port " + strconv.FormatInt(dbg.Port, 10),
+	})
+	if err != nil {
+		return fmt.Errorf("unable to start debugging: %w", err)
+	}
+
+	return nil
+}
+
+func runProgram(
+	ctx context.Context,
+	engineClient pulumirpc.EngineClient,
+	req *pulumirpc.RunRequest,
+	pwd, bin string,
+	env []string,
+) *pulumirpc.RunResponse {
+	var err error
+	var dbg *debugger
+	var cmd *exec.Cmd
+	if req.GetStartDebugger() {
+		cmd, dbg, err = debugCommand(bin)
+		if err != nil {
+			return &pulumirpc.RunResponse{
+				Error: err.Error(),
+			}
+		}
+		defer dbg.Cleanup()
+	} else {
+		cmd = exec.Command(bin)
+	}
 	cmd.Dir = pwd
-	status, err := runCmdStatus(cmd, env)
+	cmd.Env = env
+	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
+	run := func() error {
+		err := cmd.Start()
+		if err != nil {
+			return err
+		}
+		if req.GetStartDebugger() {
+			// create a sub-context to cancel the startDebugging operation when the process exits.
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+			go func() {
+				err := startDebugging(ctx, engineClient, cmd, dbg)
```

Comment:
It looks like `cmd` is unused in `startDebugging`. I think in that case we can remove the `cmd.Start()` above, and return `cmd.Run()` below, instead of splitting `Start` and `Wait` in two like this. I assume an older version of the code used `cmd.Process.Pid` or something like that.

Reply:
Yup indeed we can, thanks!